### ISSUE TITLE
Update LPWINDOWPOS flags

### ISSUE THREAD

**Issue by jmelas** [microsoft/terminal#12924]
https://github.com/microsoft/terminal/blob/41ef5554f5cb946c4a48e2b8db791265c7217fce/src/cascadia/WindowsTerminal/IslandWindow.cpp#L592

Since size is changed here shouldn't we clear the SWP_NOSIZE bit in lpwpos->flags?

**Comment by zadjii-msft** [comment]
Maybe? I'm not totally sure how those flags work. I was more under the impression that they are told to us, but aren't used in our response to the OS. Like, when that flag is set, the OS is telling us "I'm about to do something with your window, but don't worry I'm not changing your size". And our response is "okay but actually change my size". Admittedly, I don't have the strongest grasp on how user32 works so I could be mistaken. 

We can try clearing it, I'm not sure what that would change but it probably wouldn't hurt. 

@ekoschik might actually know what clearing that flag would affect, if anything.

**Comment by ekoschik** [comment]
@zadjii-msft yes, the POSCHANGING handler (if it changes the size) should be clearing the NOSIZE (and NOMOVE) flags.

Near the top of this POSCHANGING handler is a check for NOMOVE, meaning 'change the new the size and position if someone is moving the window). This should probably also check NOSIZE.

Now, if one of these flags is set but not the other, the incoming WINDOWPOS struct will have those fields NOT set. So if NOMOVE the x/y are not set (and NOSIZE means cx/cy aren't set). You could handle only cases that both move or size, or you could check for either flag being set, but know to use the window's current position or size if either of the two flags aren't set.

For example, consider someone resizing the window to move it mostly onto another monitor (do you want to detect this and 'snap' to the new monitor)?

**Comment by bfahrenfort** [comment]
@ekoschik Hi,

Here's what a potential check for `SWP_NOMOVE` and `SWP_NOSIZE` could look like. Any input?
```c++
// "if we're not moving or resizing the window, do nothing."
// GH #12924
if ( (noMove = WI_IsFlagSet(lpwpos->flags, SWP_NOMOVE)) && (noSize = WI_IsFlagSet(lpwpos->flags, SWP_NOSIZE)) )
{
      break;
}
```
Then, each boolean could be checked in this layout (avoiding additional `WI_IsFlagSet` calls):
```c++
// GH #12924: check SWP_NOMOVE and SWP_NOSIZE, clear both if either is unset
if (!noMove)
{
    lpwpos->x = newWindowRect.left;
    lpwpos->y = newWindowRect.top;
}
if (!noSize)
{
    lpwpos->cx = newWindowRect.width();
    lpwpos->cy = newWindowRect.height();
}

// at least one of the above executes, so we should clear flags.
WI_ClearFlag(lwpos->flags, SWP_NOMOVE);
WI_ClearFlag(lwpos->flags, SWP_NOSIZE);
```
Let me know what you think. I'll submit a PR after some more discussion if necessary. Thanks!

**Comment by ekoschik** [comment]
> @ekoschik Hi,
> 
> Here's what a potential check for `SWP_NOMOVE` and `SWP_NOSIZE` could look like. Any input?
> 
> ```c++
> // "if we're not moving or resizing the window, do nothing."
> // GH #12924
> if ( (noMove = WI_IsFlagSet(lpwpos->flags, SWP_NOMOVE)) && (noSize = WI_IsFlagSet(lpwpos->flags, SWP_NOSIZE)) )
> {
>       break;
> }
> ```
> 
> Then, each boolean could be checked in this layout (avoiding additional `WI_IsFlagSet` calls):
> 
> ```c++
> // GH #12924: check SWP_NOMOVE and SWP_NOSIZE, clear both if either is unset
> if (!noMove)
> {
>     lpwpos->x = newWindowRect.left;
>     lpwpos->y = newWindowRect.top;
> }
> if (!noSize)
> {
>     lpwpos->cx = newWindowRect.width();
>     lpwpos->cy = newWindowRect.height();
> }
> 
> // at least one of the above executes, so we should clear flags.
> WI_ClearFlag(lpwpos->flags, SWP_NOMOVE);
> WI_ClearFlag(lpwpos->flags, SWP_NOSIZE);
> ```
> 
> Let me know what you think. I'll submit a PR after some more discussion if necessary. Thanks!

Something seems off. The 'newWindowRect' sounds like the value from the WINDOWPOS (lpwpos). If not 'no move', the pos x/y is set (to the new window position left/top). What I think this should be doing is saying

if (noMove)
  pwpos->x = rcWIndow.left
  pwpos->y = rcWindow.top

where rcWIndow is the window's current position (GetWindowRect, or the last WINDOWPOS from WM_WINDOWPOSCHANGED).

This would produce the 'real proposed rect', which you could use to MonitorFromRect to determine which monitor the new position is on.

**Comment by bfahrenfort** [comment]
> The 'newWindowRect' sounds like the value from the WINDOWPOS (lpwpos).

Sorry I didn't clarify! `newWindowRect` contains the proposed dimensions on the proposed monitor. It's a `til::rect` retrieved from `_getQuakeModeSize`, which was called with the monitor retrieved from the values in `lpwpos`. 

My understanding is that `lpwpos` contains the proposed changes upon entry, and upon exit, the values in `lpwpos` will become the new window dimensions. Therefore, if `SWP_NOMOVE` is set, `lpwpos->x` is equal to the current window X positon, and so on.

> This would produce the 'real proposed rect'

this is done with the `RECT rcSuggested`, which is then used in getting the proposed monitor to pass to `_getQuakeModeSize`. Here's the creation of that structure:
```c++
RECT rcSuggested;
rcSuggested.left = lpwpos->x; // note: these are lpwpos's values upon entry, they have not been modified yet
rcSuggested.top = lpwpos->y;
rcSuggested.right = rcSuggested.left + lpwpos->cx;
rcSuggested.bottom = rcSuggested.top + lpwpos->cy;
```
Because `lpwpos` will contain the current window pos if unchanged, or the proposed values on a per-value basis, there's no need introduce the overhead of checking the previous position with `GetWindowRect`. 

To be honest, the guards around the `lpwpos` modifications aren't strictly necessary, they just prevent assignment if it's superfluous (if `lpwpos->x` isn't changed because `SWP_NOMOVE` is set, why would we need to assign it before exit?). At least one of those guarded blocks will execute, so like you said earlier, we must clear `SWP_NOMOVE` and SWP_NOSIZE`. Hmm, I could actually combine those two calls into a single `WI_ClearAllFlags` call...

**Comment by jmelas** [comment]
Not sure if it helps but I did something similar while implementing Per-Monitor DPI in my apps:
The intention is to resize the Window when it is moving to a different monitor.

```
        if (! (wp->flags & SWP_NOMOVE) && wp->flags & SWP_NOSIZE) //move and not size
        {
            if (new_monitor)
            {
                wp->flags &= ~SWP_NOSIZE; //force resize
                Size size = calcWindowSizeForDPI(window, dpi);
                wp->cx = size.w;
                wp->cy = size.h;
                return 0;
                }
            }
        }
```
So I am changing the size only if the user moves the window to a new monitor.

**Comment by zadjii-msft** [comment]
Hey so I maybe lost track of this - is there a practical side effect to us not doing this right? Something I could use to validate both:
1. What we're doing now causes {bug}
2. Whatever fix we try fixes {bug}

?

Seems like an easy enough fix, but I (personally) want to better understand what doing this wrong causes.

**Comment by ekoschik** [comment]
> Not sure if it helps but I did something similar while implementing Per-Monitor DPI in my apps:
> 
> ```
>         if (! (wp->flags & SWP_NOMOVE) && wp->flags & SWP_NOSIZE) //move and not size
>         {
>             if (UINT dpi = Win_GetRectDpi(Rect(wp->x, wp->y, wp->cx, wp->cy)); dpi != Win_GetWindowDpi(hWnd))
>             {
>                 wp->flags &= ~SWP_NOSIZE; //force resize
>                 Size size = calcWindowSizeForDPI(window, dpi);
>                 wp->cx = size.w;
>                 wp->cy = size.h;
>                 return 0;
>             }
>         }
> ```
> 
> So I am changing the size when a call to SetWindowPos moves the Window to a new monitor.

This will cause problems, notably when dragging the window between monitors of different DPIs.

The system expects all per-monitor DPI aware windows to always 'accept' the lParam rect in WM_DPICHANGED. Attempting to manipulate the rect in that handler OR within WM_WINDOWPOSCHANGING will cause the window to jump away from the cursor, or potentially 'wobble' (bounce back/forth between the two monitors).

Instead of this, its recommended to mark the window as 'per monitor v2' and use [WM_GETDPISCALEDSIZE](https://learn.microsoft.com/en-us/windows/win32/hidpi/wm-getdpiscaledsize). This is sent prior to sending the window a WM_DPICHANGED with the pre-scaled window size and the new DPI. For example, if the window is scaling to 200% and is moving to 300%, the window can scale the provided size from 2x to 3x, allowing it to override how it scales for DPI while still accepting the WM_DPICHANGED rect (and not flying away from the cursor).

**Comment by jmelas** [comment]
Thank you for your input! Yes indeed I am marking all windows 'per monitor v2' and I do use WM_GETDPISCALEDSIZE.
The code I posted is in WM_WINDOWPOSCHANGING. I am testing on Windows 11 and WM_WINDOWPOSCHANGING is not
called when dragging the window between monitors.
The problem this code is trying to solve is setting the correct size of the Window when I am calling SetWindowPos.
When closing I am saving the window pos to the registry and I want the window to open later at that saved position.
If this position is on a different monitor I also need to resize the window to match the size needed on that different monitor.
